internal/controllers/machine/machine_controller_phases.go

```diff
@@ -372,3 +364,30 @@ func (r *Reconciler) reconcileCertificateExpiry(ctx context.Context, _ *clusterv
 
 	return ctrl.Result{}, nil
 }
+
+// removeOutdatedOwnerRefs will remove any MachineSet or ControlPlane ownerReferences from Infrastructure Machine and
+// Bootstrap objects referred to by the machine. These ownerReferences are added on  creation as the Machine does not exist
+// when the objects are initially created.
+func removeOutdatedOwnerRefs(cluster *clusterv1.Cluster, m *clusterv1.Machine, obj *unstructured.Unstructured) {
+	cpGVK := getControlPlaneGVKForMachine(cluster, m)
+	for _, owner := range obj.GetOwnerReferences() {
+		if (owner.APIVersion == clusterv1.GroupVersion.String() && owner.Kind == "MachineSet") ||
```

Comment:
Should we only compare the Group (instead of GroupVersion)?
I think we don't know if the version has been upgraded to v1beta1 in the past, example v1alpha4 OwnerRef here: https://storage.googleapis.com/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1597204161387040768/artifacts/clusters/clusterctl-upgrade-2m76r5/resources/clusterctl-upgrade/KubeadmConfig/clusterctl-upgrade-idvv28-md-0-5pq56.yaml (clusterctl upgrade v1alpha4 test)

Reply:
Same for control plane: https://storage.googleapis.com/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1597204161387040768/artifacts/clusters/clusterctl-upgrade-2m76r5/resources/clusterctl-upgrade/KubeadmConfig/clusterctl-upgrade-idvv28-control-plane-kp9md.yaml

Reply:
Updated to remove the version comparison - tested with a v1alpha4 Cluster locally and it works as expected there. Thanks for catching this!